artiq/gateware/targets/kasli.py

```diff
@@ -367,6 +367,56 @@ def add_rtio(self, rtio_channels, sed_lanes=8):
                                                       self.get_native_sdram_if(), cpu_dw=self.cpu_dw)
         self.csr_devices.append("rtio_analyzer")
 
+    def add_eem_drtio(self, eem_drtio_channels):
+        if self.has_drtio_over_eem:
+            cdr = ClockDomainsRenamer({"rtio_rx{}".format(i): "rtio_rx{}".format(i + len(self.drtio_transceiver.channels)) 
+                for i in range(len(eem_drtio_channels))})
+
+            self.submodules.eem_transceiver = cdr(eem_serdes.EEMSerdes(self.platform, eem_drtio_channels))
+            self.csr_devices.append("eem_transceiver")
+            self.config["HAS_DRTIO_EEM"] = None
+            self.config["EEM_DRTIO_COUNT"] = len(eem_drtio_channels)
+
+            drtio_csr_group = []
+            drtioaux_csr_group = []
+            drtioaux_memory_group = []
+
+            for i in range(len(self.eem_transceiver.channels)):
+                channel = i + len(self.drtio_transceiver.channels)
+                core_name = "drtio" + str(channel)
+                coreaux_name = "drtioaux" + str(channel)
+                memory_name = "drtioaux" + str(channel) + "_mem"
+                drtio_csr_group.append(core_name)
+                drtioaux_csr_group.append(coreaux_name)
+                drtioaux_memory_group.append(memory_name)
+
+                cdr = ClockDomainsRenamer({"rtio_rx": "rtio_rx" + str(channel)})
+
+                core = cdr(DRTIOMaster(self.rtio_tsc, self.eem_transceiver.channels[i]))
+                setattr(self.submodules, core_name, core)
+                self.drtio_cri.append(core.cri)
+                self.csr_devices.append(core_name)
+
+                coreaux = cdr(DRTIOAuxController(core.link_layer, self.cpu_dw))
+                setattr(self.submodules, coreaux_name, coreaux)
+                self.csr_devices.append(coreaux_name)
+
+                memory_address = self.mem_map["drtioaux"] + 0x800*channel
+                self.add_wb_slave(memory_address, 0x800,
+                                coreaux.bus)
+                self.add_memory_region(memory_name, memory_address | self.shadow_base, 0x800)
+
+            csr_gps = self.get_csr_groups()
+            mem_gps = self.get_memory_groups()
+            for i, csr_gp in enumerate(csr_gps):
+                if csr_gp[0] == "drtio":
+                    csr_gps[i] = (csr_gp[0], csr_gp[1] + drtio_csr_group)
+                elif csr_gp[0] == "drtioaux":
+                    csr_gps[i] = (csr_gp[0], csr_gp[1] + drtioaux_csr_group)
+            for i, mem_gp in enumerate(mem_gps):
+                if mem_gp[0] == "drtioaux_mem":
+                    mem_gps[i] = (mem_gp[0], mem_gp[1] + drtioaux_memory_group)
```

Comment:
You mean to `add_csr_group("drtio", drtio_csr_group)` , `add_csr_group("drtioaux", drtioaux_csr_group)` and `add_memory_group("drtioaux_mem", drtioaux_memory_group)` only once?

I create the loop to avoid modification on `GenericMaster`
https://github.com/m-labs/artiq/blob/cd22e42cb4802d9fac9833ab747110340fc39644/artiq/gateware/targets/kasli.py#L320-L322